transport/tchannel/config.go

```diff
@@ -37,38 +42,101 @@ type InboundConfig struct{}
 
 // OutboundConfig configures a TChannel outbound.
 type OutboundConfig struct {
-	Address string `config:"address,interpolate"`
+	Choose config.PeerListConfig `config:"choose"`
 }
 
 // TransportSpec returns a TransportSpec for the TChannel unary transport. See
 // TransportConfig, InboundConfig, and OutboundConfig for details on the
 // various supported configuration parameters.
-func TransportSpec() config.TransportSpec {
+func TransportSpec(opts ...Option) config.TransportSpec {
+	// TODO: Presets. Support "with:" and allow passing those in using
+	// varargs on TransportSpec().
+	var ts transportSpec
+	for _, o := range opts {
+		switch opt := o.(type) {
+		case TransportOption:
+			ts.transportOptions = append(ts.transportOptions, opt)
+		// At present, TChannel inbound and outbound take no options, so no
+		// further types here.
+		default:
+			panic(fmt.Sprintf("unknown option of type %T: %v", o, o))
+		}
+	}
+	return ts.Spec()
+}
+
+// transportSpec holds the configurable parts of the TChannel TransportSpec.
+//
+// These are usually runtime dependencies that cannot be parsed from
+// configuration.
+type transportSpec struct {
+	transportOptions []TransportOption
+}
+
+func (ts *transportSpec) Spec() config.TransportSpec {
 	return config.TransportSpec{
 		Name:               "tchannel",
-		BuildTransport:     buildTransport,
-		BuildInbound:       buildInbound,
-		BuildUnaryOutbound: buildUnaryOutbound,
+		BuildTransport:     ts.buildTransport,
+		BuildInbound:       ts.buildInbound,
+		BuildUnaryOutbound: ts.buildUnaryOutbound,
 	}
 }
 
-func buildTransport(tc *TransportConfig, k *config.Kit) (transport.Transport, error) {
-	var opts []TransportOption
+func (ts *transportSpec) buildTransport(tc *TransportConfig, k *config.Kit) (transport.Transport, error) {
+	var config transportConfig
+	config.tracer = opentracing.GlobalTracer()
+
+	opts := ts.transportOptions
+	for _, opt := range opts {
+		opt(&config)
+	}
+
 	if tc.Address != "" {
-		opts = append(opts, ListenAddr(tc.Address))
+		config.addr = tc.Address
 	}
 	if tc.Service != "" {
-		opts = append(opts, ServiceName(tc.Service))
+		config.name = tc.Service
 	}
-	return NewTransport(opts...)
+
+	// Decide whether to use a channel-backed transport based on whether a
+	// channel instance was provided in TransportSpec options.
+	// Inbound and Outbound constructors must expect either a legacy
+	// *ChannelTransport or the *Transport with peer chooser support.
+	if config.ch != nil {
+		return config.newChannelTransport(), nil
+	}
+	return config.newTransport(), nil
 }
 
-func buildInbound(_ *InboundConfig, t transport.Transport, k *config.Kit) (transport.Inbound, error) {
+func (ts *transportSpec) buildInbound(_ *InboundConfig, t transport.Transport, k *config.Kit) (transport.Inbound, error) {
+	if x, ok := t.(*ChannelTransport); ok {
+		return x.NewInbound(), nil
+	}
 	return t.(*Transport).NewInbound(), nil
```

Comment:
We should talk about this. A policy that has made sense to me in the past is that if it's an unrecoverable error caused by a serious bug in our code, we should just panic. If this type cast fails, it's because the configuration system had a bug and fed the wrong input into `buildInbound` in which case panicking sounds pretty reasonable.